prost-derive/src/field/scalar.rs

```diff
@@ -732,3 +747,78 @@ impl quote::ToTokens for DefaultValue {
         }
     }
 }
+
+
+struct Inferred {
+    ty: Ty,
+    label: Option<Label>,
+    packed: Option<bool>,
+}
+
+impl Inferred {
+    fn new(ty: Ty, label: Option<Label>) -> Inferred {
+        let packed = match (label, &ty) {
+            (Some(Label::Repeated), &Ty::String) => Some(false),
+            (Some(Label::Repeated), &Ty::Bytes) => Some(false),
+            (Some(Label::Repeated), _) => Some(true),
+            _ => None,
+        };
+
+        Inferred{ty, label, packed}
+    }
+
+    /// Choose a sane default protobuf type from a rust type if one isn't specified
+    fn from_rust_type(ty: &Type) -> Option<Inferred> {
+        let path = match ty {
+            &Type::Path(syn::TypePath{ref path, ..}) => path,
+            _ => return None,
+        };
+        let segment = match path.segments.first() {
+            Some(syn::punctuated::Pair::End(seg)) => seg,
+            _ => return None, // NOTE: Not an inferrable scalar type
+        };
+        let label = match segment.ident.as_ref() {
+            "Option" => Some(Label::Optional),
+            "Vec" => Some(Label::Repeated),
+            _ => None,
+        };
+        let ty =
+            if label.is_some() {
+                let bracketed = match segment.arguments {
+                    syn::PathArguments::AngleBracketed(ref bracketed) => bracketed,
+                    _ => return None,
+                };
+                match bracketed.args.first() {
+                    Some(syn::punctuated::Pair::End(&syn::GenericArgument::Type(ref inner))) => inner,
+                    _ => return None,
+                }
+            } else {
+                ty
+            };
+
+        let tok = quote!(#ty);
+        if label == Some(Label::Repeated) && tok == quote!(u8) {
+            Some(Inferred::new(Ty::Bytes, None))
```

Comment:
This is also covered on line 819, are they both necessary?

Reply:
Yes.

The assignment at line ~785 (`let ty = ...`) strips either `Vec<T>` or `Option<T>` from the type causing `tok` to be just `quote!(T)`.

This check then notes, if the tok is `u8` and `Vec` was stripped from the type (e.g. if `label` is `Label::Repeated`), then we don't actually want a repeated type, we wanted protobuf `bytes` instead.

The later condition at line 819 then handles the case for `repeated bytes` and `optional bytes`.

I will try to add some comments to that effect.